x/wasm/keeper/keeper.go

```diff
@@ -483,26 +484,57 @@ func (k Keeper) migrate(
 		contractInfo.IBCPortID = ibcPort
 	}
 
-	env := types.NewEnv(sdkCtx, contractAddress)
+	var response *wasmvmtypes.Response
 
-	// prepare querier
-	querier := k.newQueryHandler(sdkCtx, contractAddress)
-
-	prefixStoreKey := types.GetContractStorePrefix(contractAddress)
-	vmStore := types.NewStoreAdapter(prefix.NewStore(runtime.KVStoreAdapter(k.storeService.OpenKVStore(sdkCtx)), prefixStoreKey))
-	gasLeft := k.runtimeGasForContract(sdkCtx)
-	res, gasUsed, err := k.wasmVM.Migrate(newCodeInfo.CodeHash, env, msg, vmStore, cosmwasmAPI, &querier, k.gasMeter(sdkCtx), gasLeft, costJSONDeserialization)
-	k.consumeRuntimeGas(sdkCtx, gasUsed)
+	// check for migrate version
+	oldCodeInfo := k.GetCodeInfo(ctx, contractInfo.CodeID)
+	oldReport, err := k.wasmVM.AnalyzeCode(oldCodeInfo.CodeHash)
 	if err != nil {
 		return nil, errorsmod.Wrap(types.ErrVMError, err.Error())
 	}
-	if res == nil {
-		// If this gets executed, that's a bug in wasmvm
-		return nil, errorsmod.Wrap(types.ErrVMError, "internal wasmvm error")
-	}
-	if res.Err != "" {
-		return nil, types.MarkErrorDeterministic(errorsmod.Wrap(types.ErrMigrationFailed, res.Err))
+
+	switch {
+	case report.ContractMigrateVersion != nil && oldReport.ContractMigrateVersion != nil:
+		// if both are set, we only migrate if the new version is higher
+		switch {
+		case *report.ContractMigrateVersion < *oldReport.ContractMigrateVersion:
+			// prevent downgrades
+			return nil, errorsmod.Wrapf(
+				types.ErrMigrationFailed,
+				"new migrate version %d is lower than current version %d",
+				*report.ContractMigrateVersion,
+				*oldReport.ContractMigrateVersion,
+			)
+		case *report.ContractMigrateVersion > *oldReport.ContractMigrateVersion:
+			// in the future, we will provide MigrateInfo to the contract in this case
+			response, err = k.callMigrateEntrypoint(sdkCtx, contractAddress, wasmvmtypes.Checksum(newCodeInfo.CodeHash), msg, setupCost)
+			if err != nil {
+				return nil, err
+			}
+		}
+		// no need to call the migrate entrypoint if the versions are the same
+	case report.ContractMigrateVersion != nil:
+		// migrating from contract without migrate version to one with migrate version
+		// in the future, we will provide MigrateInfo to the contract in this case
+		response, err = k.callMigrateEntrypoint(sdkCtx, contractAddress, wasmvmtypes.Checksum(newCodeInfo.CodeHash), msg, setupCost)
+		if err != nil {
+			return nil, err
+		}
+	case oldReport.ContractMigrateVersion != nil:
+		// prevent migrating from contract with migrate version to one without migrate version
```

Comment:
This one might be debatable. Should we allow it?

Reply:
What about this: the default case is that migrate is called. Only if both migrate versions are set and they are equal, migrate is skipped. That pretty much keeps the current behaviour except for one simple optimization.

I would not differentiate between upgrading and downgrading. The contract developer can handle this themselves.

Reply:
Ok, that simplifies it a lot.